File: apps/server/src/supervised/runtime/kernel_host.py
import contextlib
import io
import json
import sys
import traceback

state = {"__builtins__": __builtins__}

for raw_line in sys.stdin:
    request = None
    try:
        request = json.loads(raw_line)
        local_scope = {"state": state, "input": request.get("input"), "result": None}
        output = io.StringIO()
        with contextlib.redirect_stdout(output):
            exec(request["code"], state, local_scope)
        response = {
            "id": request["id"],
            "ok": True,
            "result": local_scope.get("result"),
            "stdout": output.getvalue(),
        }
    except Exception as error:
        response = {
            "id": request.get("id", "unknown") if isinstance(request, dict) else "unknown",
            "ok": False,
            "error": str(error),
            "trace": traceback.format_exc(limit=4),
        }
    sys.stdout.write(json.dumps(response, separators=(",", ":")) + "\n")
    sys.stdout.flush()

File: apps/server/src/supervised/runtime/kernelHost.mjs
import readline from "node:readline";

const state = Object.create(null);
const AsyncFunction = Object.getPrototypeOf(async function () {}).constructor;
const lines = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });

for await (const line of lines) {
  let request;
  try {
    request = JSON.parse(line);
    const fn = new AsyncFunction(
      "state",
      "input",
      `"use strict";\n${request.code}\n`,
    );
    const result = await fn(state, request.input ?? null);
    process.stdout.write(`${JSON.stringify({ id: request.id, ok: true, result })}\n`);
  } catch (error) {
    process.stdout.write(
      `${JSON.stringify({
        id: request?.id ?? "unknown",
        ok: false,
        error: error instanceof Error ? error.message : String(error),
      })}\n`,
    );
  }
}
